### PR TITLE
Support for hooks on bastion instance groups

### PR DESCRIPTION
**Pull request by Cryptophobia**
This PR is supposed to close https://github.com/kubernetes/kops/issues/3907 but not sure if this is the only place to touch. Need feedback if on the right track @justinsb @chrislovecnm .

Docker-engine is already enabled and working because user-data support and nodeup.sh were added in https://github.com/kubernetes/kops/pull/4452 .

### CODE REVIEW

**Comment by Cryptophobia** [comment]
assign @andrewsykim

**Comment by Cryptophobia** [comment]
@idealhack  Code has been fixed and test pass now. Would it be nice to write an e2e test for bastion hooks. Where would be the best place to write a integration systems test? Can someone point me to a location or file?

**Comment by idealhack** [comment]
@Cryptophobia I’m not sure about that too, defer to kops maintainers.

File: pkg/model/bootstrapscript.go [hunk 1]
```diff
@@ -98,7 +98,16 @@ func (b *BootstrapScript) buildEnvironmentVariables(cluster *kops.Cluster) (map[
 // template file, substituting in specific env vars & cluster spec configuration
 func (b *BootstrapScript) ResourceNodeUp(ig *kops.InstanceGroup, cluster *kops.Cluster) (*fi.ResourceHolder, error) {
 	// Bastions can have AdditionalUserData, but if there isn't any skip this part
-	if ig.IsBastion() && len(ig.Spec.AdditionalUserData) == 0 {
+	// Check if bastion has hooks enabled, in that special case we want to set hooks up
+	var bastionHooks []kops.HookSpec
+	var err error
+	if ig.IsBastion() {
+		bastionHooks, err = b.getRelevantHooks(ig.Spec.Hooks, ig.Spec.Role)
```
Comment (justinsb) [inline pkg/model/bootstrapscript.go]: We should check err here

File: pkg/model/bootstrapscript.go [hunk 2]
```diff
@@ -98,7 +98,16 @@ func (b *BootstrapScript) buildEnvironmentVariables(cluster *kops.Cluster) (map[
...
+	} else {
+		//Do nothing as bastionHooks should be empty
+	}
+
+	if ig.IsBastion() && len(ig.Spec.AdditionalUserData) == 0 && !(len(bastionHooks) > 0) {
```
Comment (justinsb) [inline pkg/model/bootstrapscript.go]: Probably easier on that last term to say `.. && len(bastionHooks) == 0`

File: pkg/model/bootstrapscript.go (hunk 1 above)
Comment (Cryptophobia) [reply pkg/model/bootstrapscript.go]: Fixed in https://github.com/kubernetes/kops/pull/5856/commits/682c5426023c70367628ad08f157fec0577de6b3

File: pkg/model/bootstrapscript.go (hunk 2 above)
Comment (Cryptophobia) [reply pkg/model/bootstrapscript.go]: Fixed in https://github.com/kubernetes/kops/pull/5856/commits/682c5426023c70367628ad08f157fec0577de6b3

**Comment by Cryptophobia** [comment]
> But to make that decision - what is the use-case here? I'm guessing you want to set up something like a VPN or SSH keys, in which case I'm guessing "cluster level hooks don't work on bastions unless you explicitly specify a role" will be acceptable (without the field, probably with the warning), and you don't want to run e.g. kubelet on the nodes.

I think the "cluster level hooks don't work on bastions unless you explicitly specify a role" would not be breaking change and will be satisfactory. We ran into this issue when we wanted to install `netcat` and networking tools to create tunnels and execute remote commands through the bastions. Seemed like networking tools that we needed were not available on the official kops Debian images. I see multiple use cases but ours was mainly for networking tools and VPN/SSH tunnels like you said.

> Two small code nits

Addressed in https://github.com/kubernetes/kops/pull/5856/commits/682c5426023c70367628ad08f157fec0577de6b3

> I suspect we want to ensure that most of those are no-ops on bastions. For example, we probably don't want to install Kubelet (?).

I agree with this. There are lots of things that we normally install and maybe would want to skip for bastions.

**Comment by tsuna** [comment]
I need this too. Can we push this through the finish line? :)

**Comment by tsuna** [comment]
Actually this quite fix the issue in #5242 as it seems that this is adding just support for defining hooks in the ig of the bastions. The cluster-wide fileAssets and hooks are still ignored for bastions from what I can tell.

**Comment by Cryptophobia** [comment]
> Actually this quite fix the issue in #5242 as it seems that this is adding just support for defining hooks in the ig of the bastions. The cluster-wide fileAssets and hooks are still ignored for bastions from what I can tell.

This is probably for the best intention. There are lots of processes (like kubelet...) that run on the nodes but do not run in the bastion instances, refer to @justinsb's comment https://github.com/kubernetes/kops/pull/5856#issuecomment-439688913.

**Comment by tsuna** [comment]
> I can push this through if we make a decision on how best to handle the upgrade/rollout process. We don't want to break existing clusters. Also need someone to point to what other files changes need to be made.

Maybe the default behavior can remain the same but we add a knob in the ig that can be turned to say "run the hooks for this ig".

**Comment by justinsb** [comment]
Thanks for the PR @Cryptophobia 

We discussed in [office-hours today](https://www.youtube.com/watch?v=SdW4Y7FgZfI&index=2&list=PL69nYSiGNLP2NcuGKoAzme6ZtZmrJlsaR&t=930s) the need to prioritize bug fixes until we can get 1.12 out (which will hopefully mean we get to everything faster overall), and I believe this is a feature, so I'm going to move this to the next release.

/kind feature
/milestone 1.13

**Comment by mazzy89** [comment]
Thank you @rifelpet for opening this.

I would be interested to push this out. How can we achieve that? I assume @Cryptophobia you step aside, don't you?

**Comment by Cryptophobia** [comment]
@mazzy89 , if I remember correctly this was stuck in limbo because was not sure what other files needed to be changed to allow for hooks to be ran on bastions without breaking existing clusters.

Basically there needs to be a knob somewhere that ignores typical hooks that install things like `kubelet` and the other default hooks for bastions...

**Comment by hakman** [comment]
@mazzy89 if you need any help, please let me know.

**Comment by mazzy89** [comment]
@hakman thank you for the follow up here.

Actually we are going to move out of the bastion host in favor of a dedicated instance managed by Terraform. This would avoid us to rollout changes to this instance easily, more often and not tightened to kops - which anyway produce a Terraform output so keep us always close.

If there is not much interest here then we can close this.

**Comment by hakman** [comment]
Thanks, @mazzy89. Closing for now. If anyone is interested about this, can be easily reopened.

**Comment by estahn** [comment]
@hakman If this could be reopened that would be great. We're trying to install AWS Inspector via hooks on the bastion.

**Comment by hakman** [comment]
@estahn If anyone wants to work on this, they should create a new PR, but this is not a trivial task.
In the mean time, I would suggest to try using `additionalUserData `, example:
https://kops.sigs.k8s.io/operations/images/#debian-10-buster

**Pull request closed without merging**